### ISSUE TITLE
Frozen Copy of Data from Prod to Test Against in Dev

### ISSUE THREAD

**Issue by kells1986** [Passiolife/passio_devops#43]
One of the issues that I'm having with the ML testing at the moment is the dynamic changes to labels and hierarchy structure that are being applied by the data team in prod. Having defined a list of labels to train on, after a day or so some of those labels may no longer exist, or their hierarchical relationship may have changed.

To remedy this we'd like to have a frozen copy of all of the data from prod, and an instance of the API so that the ML code can run tests against that static data.

I don't know what the best / easiest / most efficient way of doing this is, so I'm open to suggestions. My thoughts were:
* A new GCP project passio-test-env with the database, buckets and API setup
* A new database within dev, and a new api url within dev that points to the frozen database

- [ ] decide on a path forward
- [ ] make a copy of the prod data to the selected frozen location (maybe make this a cron job?)
- [ ] deploy the API to serve the frozen data.
- [ ] @drstarson please review
- [ ] @Loenike please review

**Comment by ncheikh-passio** [comment]
This definitely fits with the discussion on copying data from Prod -> Dev that was had a while ago.  If we needed another environment that was less dynamic, it would be simple to have multiple copy/migration jobs at desired frequencies. For an ml testing environment we could simply decrease the sync frequency with prod as needed.

**Comment by sptrakesh** [comment]
@kells1986 setting up a frozen environment should be easy enough.  The only real question is about updating the database.  Once updated you will run into the same issue that you have at present with the *live* database.

**Comment by Loenike** [comment]
I think we should move forward with this. Let's be clear though for what exactly we want to use this frozen environment. If we just want to use the frozen environment so that James has a static environment for doing the ML training, then data-wise I think we only need to do the following:

- rsync the 2 buckets with assets and cropped-assets
- delete/rewrite or update all database entries from the frozen environment

@sptrakesh Would this resolve the issue that you mentioned?

> @kells1986 setting up a frozen environment should be easy enough. The only real question is about updating the database. Once updated you will run into the same issue that you have at present with the live database.

**Comment by sptrakesh** [comment]
> @sptrakesh Would this resolve the issue that you mentioned?

I do not think so.  If the current issue is that labels have changed, then it will manifest itself anytime we re-sync the training environment with the production database (regardless of whether it is a drop and replace, or merge).

**Comment by Loenike** [comment]
@kells1986 @sptrakesh I might misunderstand this, but I guess, that the issue is that James needs a frozen environment, so that he can run several training iterations on the SAME data timestamp.

The issue for James is more that the production database is constantly being modified, which means that if he wants to repeat a training (because there was an issue in his code or whatever) - he cannot do that, because the production database has already changed since that moment.

**Comment by sptrakesh** [comment]
> The issue for James is more that the production database is constantly being modified, which means that if he wants to repeat a training (because there was an issue in his code or whatever) - he cannot do that, because the production database has already changed since that moment.

If that is the case, then a *training* environment where we sync production database on a set interval (1 week?) can be set up.

**Comment by Loenike** [comment]
Yes, @kells1986 pleas confirm, but I guess this is exactly what we would need here.

**Comment by kells1986** [comment]
>The issue for James is more that the production database is constantly being modified, which means that if he wants to repeat a training (because there was an issue in his code or whatever) - he cannot do that, because the production database has already changed since that moment.

Yes, this is the issue, for debugging purposes it's hard (impossible) to rerun the same training on the labels I've been handed first time around because the data is constantly changing.

If we can do the sync say once a week we can train on that data repeatably in that time frame incase there are any errors in the pipeline, or I want to test a new update.

>If that is the case, then a training environment where we sync production database on a set interval (1 week?) can be set up.

Yes, that would be great.

>Yes, @kells1986 please confirm, but I guess this is exactly what we would need here.
@kells1986

Exactly, confirmed.

**Comment by Loenike** [comment]
@sptrakesh Are you blocked by anything on this? If no, I think we should move forward with this asap.

**Comment by Loenike** [comment]
@drstarson Just FYI

**Comment by sptrakesh** [comment]
> @sptrakesh Are you blocked by anything on this? If no, I think we should move forward with this asap.

No.

**Comment by kells1986** [comment]
@sptrakesh please let me know if there are any tasks you can hand off to me to help out with the setup

**Comment by sptrakesh** [comment]
Deployed internal only api stack to new `passio-ml-training-env`.  There should be no need for external facing load balancer for this environment.

**Comment by sptrakesh** [comment]
Set up cron job to export production database every Sunday at 03:00 hours UTC.  Set up cron job to import production database export file every Sunday at 03:30 hours UTC on the training environment.

**Comment by kells1986** [comment]
After talking with @Loenike I think we can skip the stage of copying the buckets. 

We now have the database and API frozen, and the assets and regions don't get deleted, and are stored by the UUID. With the frozen database we can get the label for each region or asset at the time of the database import.

We don't yet do any deleting of assets, but if we were to do that in the future I could make my code tolerant to missing files.

I've just granted the service account `11320877205-compute@developer.gserviceaccount.com` (ml-training account) the role Storage Object Viewer, which means I have read-only permissions.

**Comment by drstarson** [comment]
After I moved the project to our correct billing account some of the quotas got adjusted up, but not all. In particular we are running into one backend quota at this time. I requested increase. They are typically fast. I will keep track of the issue.

**Comment by kells1986** [comment]
Is it possible to do a bulk request of quotas to match that of the prod env? 

There are an enormous number of quotas within the project and the vast majority are lower then the prod env. If we keep running into these issues one-by-one then it's going to be an enormous time sink.

If it's not possible to do a bulk request we might have to weigh up the time it will take to fix each one individually, vs the time it would take to setup a new project with quotas matching the prod env.

**Comment by kells1986** [comment]
I've setup the environment to allow training of Object Detection, and I'm trying to test. However, object detection data prep is very demanding on the API.

It seems like I'm getting a lot of 502 errors, even after waiting some time to allow the API to scale up.

@sptrakesh does the API on the ml-training env have the same capacity as the prod env?

**Comment by sptrakesh** [comment]
Yes, the API cluster is set up identical to prod.  I do not see huge database spikes, but the database is much smaller capacity than prod.

**Comment by kells1986** [comment]
I just tried again and got a lot further before I hit the 502, I keep seeing this message: `https://api-training.passiolife.com//v2/assets_api/regions/by/filter/`

**Comment by sptrakesh** [comment]
This makes a bit more sense now.  Database was maxed out for some period of time.

https://console.cloud.google.com/sql/instances/passio/overview?project=passio-ml-training-env

**Comment by kells1986** [comment]
OK, can we scale up the DB to resolve the issue?

**Comment by sptrakesh** [comment]
@drstarson I configured the database with 6 CPU.  Do we want to bump this up to 16 like currently on prod?  If ML never uses prod, we can scale that down to the previous 4, since we no longer have the very heavy hierarchy implementation.

**Comment by sptrakesh** [comment]
General note.  Eventually we may want to have the full `ml-training` environment automated, where we spin it up, copy the database, and then kick off ML jobs.  At the end, we destroy the whole environment.

**Comment by kells1986** [comment]
Still seeing quite a lot of 502 errors: `ERROR:tensorflow:('https://api-training.passiolife.com//v2/assets_api/assets/id/0ecfe75a-c614-4342-9ab7-92a5b6c79be3/',)`. The request rate should have dropped off now, past the initial spike, and it doesn't look like the database CPU usage is very high

**Comment by drstarson** [comment]
@kells1986 @sptrakesh is this resolved now?

**Comment by sptrakesh** [comment]
I updated the GKE cluster with bigger nodes last night, let us see if that makes a difference.  After the nutrition API is in place, we can get back to looking at adding some API calls that are designed specifically for the ML jobs.

**Comment by kells1986** [comment]
Looks like the 502 errors are resolved. I've just run quite a lot of queries at the same time (training for object detection and HNN), and not seen any errors.

I also increased the amount of time to wait if there is a 502.

Also looks like the database update last night all went smoothly.
